src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs

```diff
@@ -102,18 +100,14 @@ public override bool GetMemberAndWriteJson(object obj, ref WriteStack state, Utf
             bool success;
             T value = Get!(obj);
 
-            if (value == null)
+            if (IgnoreDefaultValuesOnWrite && (
+                default(T) is null ? value is null : EqualityComparer<T>.Default.Equals(default, value)))
```

Comment:
@layomia This one should be changed to the following when you start working on defaults:

```csharp
DefaultValue is null
    ? value is null // fast check for reference types in case when the default is null
    : EqualityComparer<T>.Default.Equals(DefaultValue, value) // fast for non-shared generics, but slow in other cases
```